src/object.c

```diff
@@ -580,27 +580,35 @@ void freeStreamObject(robj *o) {
 }
 
 void incrRefCount(robj *o) {
-    if (o->refcount < OBJ_FIRST_SPECIAL_REFCOUNT) {
-        o->refcount++;
+    unsigned short refcount;
+    atomicGet(o->refcount, refcount);
+
+    if (refcount < OBJ_FIRST_SPECIAL_REFCOUNT) {
+        serverAssert(refcount != OBJ_FIRST_SPECIAL_REFCOUNT - 1);
+        atomicIncr(o->refcount, 1);
     } else {
-        if (o->refcount == OBJ_SHARED_REFCOUNT) {
+        if (refcount == OBJ_SHARED_REFCOUNT) {
             /* Nothing to do: this refcount is immutable. */
-        } else if (o->refcount == OBJ_STATIC_REFCOUNT) {
+        } else if (refcount == OBJ_STATIC_REFCOUNT) {
             serverPanic("You tried to retain an object allocated in the stack");
         }
     }
 }
 
 void decrRefCount(robj *o) {
-    if (o->refcount == OBJ_SHARED_REFCOUNT)
+    unsigned short refcount;
+    atomicGet(o->refcount, refcount);
+
+    if (refcount == OBJ_SHARED_REFCOUNT)
         return; /* Nothing to do: this refcount is immutable. */
 
-    if (unlikely(o->refcount <= 0)) {
+    if (unlikely(refcount <= 0)) {
         serverPanic("illegal decrRefCount for object with: type %u, encoding %u, refcount %d",
-            o->type, o->encoding, o->refcount);
+            o->type, o->encoding, refcount);
     }
 
-    if (--(o->refcount) == 0) {
+    atomicDecr(o->refcount, 1);
```

Comment:
I think we need to use the returned value here to avoid double free due to race.
Say if `o->refcount == 1`, two threads enter `decrRefCount()` and `atomicGet()`  `refcount == 1`, would double free (though two `decrRefCount()`s with refcount of 1 is def a bug).

```suggestion
    refcount = atomicDecr(o->refcount, 1);
```

Reply:
good catch.

Reply:
done with https://github.com/redis/redis/pull/14608/commits/517efcca90d8f1f4cc21580304534639297bcf5b